File: public_web/src/lib/vixrexDecisionContract.ts
import type { VixrexMatchClass } from "./vixrexIntentResolver";

export const VIXREX_DECISION_CONTRACT_VERSION = 1 as const;

export type VixrexDecisionKind =
  | "not_understood"
  | "needs_clarification"
  | "needs_special_flow"
  | "validated_action"
  | "validated_action_group"
  | "blocked";

/**
 * `pending_slot`, iki turlu konuşmada alan önceki mesajda kesin olarak
 * seçilmişken ikinci mesaj yalnız değeri taşıdığında kullanılır. Bu bir
 * matcher skoru değildir; sahte bir `exact_*` sonucu üretmemek için açıkça
 * ayrı tutulur. 5.4 pending envelope bunu kanonik state'e taşıyacaktır.
 */
export type VixrexActionMatchClass = VixrexMatchClass | "pending_slot";

export interface VixrexValidatedAction {
  contractVersion: typeof VIXREX_DECISION_CONTRACT_VERSION;
  domain: "storefront";
  actionType: "set_field";
  fieldKey: string;
  normalizedValue: string | number | boolean | null;
  matchClass: VixrexActionMatchClass;
}

export function createValidatedAction(params: {
  fieldKey: string;
  normalizedValue: unknown;
  matchClass: VixrexActionMatchClass;
}): VixrexValidatedAction {
  const value = params.normalizedValue;
  if (
    value !== null &&
    typeof value !== "string" &&
    typeof value !== "number" &&
    typeof value !== "boolean"
  ) {
    throw new TypeError("validated_action normalizedValue desteklenmeyen tipte.");
  }

  return {
    contractVersion: VIXREX_DECISION_CONTRACT_VERSION,
    domain: "storefront",
    actionType: "set_field",
    fieldKey: params.fieldKey,
    normalizedValue: value,
    matchClass: params.matchClass,
  };
}

File: lib/services/vixrex_nlu/vixrex_decision_contract.dart
/// Vixrex Akıllı Motor decision/action sözleşmesi.
///
/// Decision katmanı yalnız bu sonucu üretir; persistence yapmaz ve
/// "kaydedildi" başarı semantiği taşımaz.
abstract final class VixrexDecisionKind {
  static const notUnderstood = 'not_understood';
  static const needsClarification = 'needs_clarification';
  static const needsSpecialFlow = 'needs_special_flow';
  static const validatedAction = 'validated_action';
  static const validatedActionGroup = 'validated_action_group';
  static const blocked = 'blocked';
}

class VixrexValidatedAction {
  static const int currentContractVersion = 1;

  final int contractVersion;
  final String domain;
  final String actionType;
  final String fieldKey;
  final Object? normalizedValue;
  final String matchClass;

  const VixrexValidatedAction({
    this.contractVersion = currentContractVersion,
    this.domain = 'storefront',
    this.actionType = 'set_field',
    required this.fieldKey,
    required this.normalizedValue,
    required this.matchClass,
  }) : assert(
         normalizedValue == null ||
             normalizedValue is String ||
             normalizedValue is num ||
             normalizedValue is bool,
         'validated_action normalizedValue desteklenmeyen tipte.',
       );
}
